lib/oci/manifest.go

```diff
@@ -0,0 +1,283 @@
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/containers/build/util"
+
+	//specs "github.com/opencontainers/image-spec/specs-go"
+	ociImage "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+var (
+	// ErrNotFound is returned when acbuild is asked to remove an element from a
+	// list and the element is not present in the list
+	ErrNotFound = fmt.Errorf("element to be removed does not exist in this image")
+)
+
+// Manifest is a struct with an open handle to a manifest that it can manipulate
+type Image struct {
+	ociPath  string
+	refName  string
+	config   ociImage.Image
+	manifest ociImage.Manifest
+	ref      ociImage.Descriptor
+}
+
+func LoadImage(ociPath string) (*Image, error) {
+	i := &Image{
+		ociPath: ociPath,
+		refName: "latest",
+	}
+
+	refDir := path.Join(ociPath, "refs")
+	blobDir := path.Join(ociPath, "blobs")
+
+	// Look for refs
+	refFileInfos, err := ioutil.ReadDir(refDir)
+	if err != nil {
+		return nil, err
+	}
+	if len(refFileInfos) == 0 {
+		return nil, fmt.Errorf("no refs found in image")
+	}
+	// We need to pick a ref, if there's more than one we don't know which one
+	// the user wishes to modify. Let's just pick the first one.
+	i.refName = path.Base(refFileInfos[0].Name())
+
+	// Open the ref file, read it, unmarshal it, and parse the manifest's
+	// hash
+	refFile, err := os.OpenFile(path.Join(refDir, i.refName), os.O_RDWR, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer refFile.Close()
+	refBlob, err := ioutil.ReadAll(refFile)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(refBlob, &i.ref)
+	if err != nil {
+		return nil, err
+	}
+	manifestHashAlgo, manifestHash, err := splitHash(i.ref.Digest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Open the manifest, read it, unmarshal it, and parse the config's hash
+	manifestFile, err := os.OpenFile(path.Join(blobDir, manifestHashAlgo, manifestHash), os.O_RDWR, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer manifestFile.Close()
+	manifestBlob, err := ioutil.ReadAll(manifestFile)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(manifestBlob, &i.manifest)
+	if err != nil {
+		return nil, err
+	}
+	configHashAlgo, configHash, err := splitHash(i.manifest.Config.Digest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Open the config, read it, unmarshal it
+	configFile, err := os.OpenFile(path.Join(blobDir, configHashAlgo, configHash), os.O_RDWR, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer configFile.Close()
+	configBlob, err := ioutil.ReadAll(configFile)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(configBlob, &i.config)
+	if err != nil {
+		return nil, err
+	}
+
+	return i, nil
+}
+
+func splitHash(hash string) (string, string, error) {
+	pieces := strings.Split(hash, ":")
+	if len(pieces) != 2 {
+		return "", "", fmt.Errorf("error parsing digest in image ref: %q", hash)
+	}
+	return pieces[0], pieces[1], nil
+}
+
+func (i *Image) Close() error {
+	return nil
+}
```

Comment:
It's for the appc implementation of the `lib.Manifest` interface, which keeps an open handle to its manifest, but now that you've pointed this out I've noticed that this never gets called. I'm going to change things around so it's not needed, which will probably be easier than updating everywhere in the acbuild package to call Close.